src/inlinestrings.jl

```diff
@@ -0,0 +1,770 @@
+import Base: ==
+
+export InlineString, InlineStringType
+
+abstract type InlineString <: AbstractString end
+
+for sz in (1, 4, 8, 16, 32, 64, 128, 256)
+    nm = Symbol(:InlineString, max(1, sz - 1))
+    @eval begin
+        primitive type $nm <: InlineString $(sz * 8) end
+        export $nm
+    end
+end
+
+_bswap(x::T) where {T <: InlineString} = T === InlineString1 ? x : Base.bswap_int(x)
+
+const InlineStrings = Union{InlineString1,
+                            InlineString3,
+                            InlineString7,
+                            InlineString15,
+                            InlineString31,
+                            InlineString63,
+                            InlineString127,
+                            InlineString255}
+
+function Base.promote_type(::Type{T}, ::Type{S}) where {T <: InlineString, S <: InlineString}
+    T === InlineString1 && return S
+    S === InlineString1 && return T
+    T === InlineString3 && return S
+    S === InlineString3 && return T
+    T === InlineString7 && return S
+    S === InlineString7 && return T
+    T === InlineString15 && return S
+    S === InlineString15 && return T
+    T === InlineString31 && return S
+    S === InlineString31 && return T
+    T === InlineString63 && return S
+    S === InlineString63 && return T
+    T === InlineString127 && return S
+    S === InlineString127 && return T
+    return InlineString255
+end
+
+Base.ncodeunits(::InlineString1) = 1
+Base.ncodeunits(x::InlineString) = Int(Base.trunc_int(UInt8, x))
+Base.codeunit(::InlineString) = UInt8
+
+Base.@propagate_inbounds function Base.codeunit(x::T, i::Int) where {T <: InlineString}
+    @boundscheck checkbounds(Bool, x, i) || throw(BoundsError(x, i))
+    if T === InlineString1
+        return Base.bitcast(UInt8, x)
+    else
+        return Base.trunc_int(UInt8, Base.lshr_int(x, 8 * (sizeof(T) - i)))
+    end
+end
+
+function Base.String(x::T) where {T <: InlineString}
+    len = ncodeunits(x)
+    out = Base._string_n(len)
+    if T === InlineString1
+        GC.@preserve out unsafe_store!(pointer(out), codeunit(x, 1))
+        return out
+    end
+    ref = Ref{T}(_bswap(x))
+    GC.@preserve ref out begin
+        ptr = convert(Ptr{UInt8}, Base.unsafe_convert(Ptr{T}, ref))
+        unsafe_copyto!(pointer(out), ptr, len)
+    end
+    return out
+end
+
+function Base.Symbol(x::T) where {T <: InlineString}
+    ref = Ref{T}(_bswap(x))
+    return ccall(:jl_symbol_n, Ref{Symbol},
+        (Ref{T}, Int), ref, sizeof(x))
+end
+
+# add a codeunit to end of string method
+function addcodeunit(x::T, b::UInt8) where {T <: InlineString}
+    if T === InlineString1
+        return x, false
+    end
+    len = ncodeunits(x)
+    (len + 1) == sizeof(T) && return x, true
```

Comment:
what the second returned value mean? As it seems `InlineString1` and other types of `InlineString*` are not handled consistently?

Reply:
2nd returned value is `overflowed`. Down in the `Parsers.xparse` code, we need to track if we overflow a given inline string type so we know whether we need to promote. This is similar to how we do it for `Int` parsing.